### PR TITLE
Minimally-invasive separation of bitcoin keys from ECDSA signature types

### PR DESCRIPTION
**Pull request by dr-orlovsky**
This PR tries to do a minimally-invazive separation of signature- and key-related types, previously mixed in a single `util::ecdsa` module.

Rationale: bitcoin key types are not specific for signature algorithm. See discussion at #588.

This PR became possible after we moved on new `secp256k1` version exposing `XonlyPublicKey` type, since now all key types may co-exist in a single module under different names

The PR goal is achieved through
- Renaming ecdsa mod into private ec module such that the code is not copied and diff size is small;
- Introducing dummy ecdsa mod back in the next commit and re-exporiting only signature types from internal `ec` mod in it;
- Re-exporting all key types under `key` module, removing previous depreciation message for bitcoin keys.

### CODE REVIEW

**Comment by Kixunil** [comment]
Concept ACK. Given how trivial the `ecdsa` mod is It think this would be better without eedab33955840527cee1f05b4e9cc7e4ec674c6c

Was also thinking of moving things outside of `util` while you're at it but maybe it's too unrelated.

File: src/util/schnorr.rs [hunk 1]
```diff
@@ -20,6 +20,7 @@
 use core::fmt;
 use prelude::*;
 
+#[deprecated(since = "0.27.0", note = "Please use `util::key` instead")]
```
Comment (apoelstra) [inline src/util/schnorr.rs:23]: In 426edd638f143532bd19573344dddbd499810b61:

ITYM since 0.28.0

File: src/util/ec.rs [hunk 2]
```diff
@@ -11,11 +11,9 @@
 // If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 //
 
-//! ECDSA Bitcoin keys.
-//!
-//! This module provides ECDSA keys used in Bitcoin that can be roundtrip
-//! (de)serialized.
+//! Elliptic curve Bitcoin keys and signatures.
 //!
+//! This module provides compressed and uncompressed Bitcoin keys and ECDSA signature-related types.
```
Comment (apoelstra) [inline src/util/ec.rs:16]: In f94fb6dde81e28aacd268b96d49ae30a9427c664:

I don't like this rename. The implication is that xonly keys are not EC keys. They should be called "full keys" or something like that. Honestly I think the entire contents of util/ecdsa.rs could just be pulled into key.rs directly.

File: src/lib.rs [hunk 3]
```diff
@@ -137,10 +137,7 @@ pub use util::sighash::SchnorrSigHashType;
 
 pub use util::ecdsa::{self, EcdsaSig, EcdsaSigError};
 pub use util::schnorr::{self, SchnorrSig, SchnorrSigError};
-#[deprecated(since = "0.26.1", note = "Please use `ecdsa::PrivateKey` instead")]
-pub use util::ecdsa::PrivateKey;
-#[deprecated(since = "0.26.1", note = "Please use `ecdsa::PublicKey` instead")]
-pub use util::ecdsa::PublicKey;
```
Comment (sanket1729) [inline src/lib.rs:143]: FYI, for the future, this warning does not really work. If you want to deprecate a type, you would have to have to 
- Rename the existing type
- Create a new type `pub type OldName = NewName`
- Add deprecated above it

**Comment by sanket1729** [comment]
> Honestly I think the entire contents of util/ecdsa.rs could just be pulled into key.rs directly.

Maybe we can punt on re-organizing keys to #751 that would massive crate re-organization anyways? And I think this is not critical for taproot release.  
As mentioned in the review, using `bitcoin::PublicKey` did not generate any deprecation downstream. I think we can get away with nuking `ecdsa::PublicKey` and `ecdsa::PrivateKey` directly.

**Review by sanket1729 — COMMENTED** [review]
As mentioned in the review comment, we did not really deprecate bitcoin::PublicKey and bitcoin::PrivateKey. I think we can get away with nuking `ecdsa::PublicKey` and `ecdsa::SecretKey`. 

And I think renaming and moving things can be left to #751 (which I have not reviewed, but might be a better place to organize things in the library). 

I think we can limit the PR to removing the `ecdsa::PublicKey` and `ecdsa::PrivateKey` and leave the organization and signatures and keys to a longer-term project. 

We probably want key stuff in `util::key` and leave only the signature stuff in ecdsa.rs. I am also okay with punting on this till 0.29 as I don't think this is critical for taproot release.

**Comment by dr-orlovsky** [comment]
I've redone this PR trying to follow your suggestions. Now it effectively moves keys from `ecdsa` mod to `key` mod - and it is done in a way that you can check that the code has not changed by reviewing on a per-commit basis.

File: src/util/schnorr.rs [hunk 4]
```diff
@@ -20,6 +20,7 @@
 use core::fmt;
 use prelude::*;
 
+#[deprecated(since = "0.28.0", note = "Please use `util::key` instead")]
 pub use secp256k1::{XOnlyPublicKey, KeyPair};
```
Comment (Kixunil) [inline src/util/schnorr.rs:24]: Need to use `type` as @sanket1729 said above.

**Comment by apoelstra** [comment]
ACK 6ecc8809112ef17c428c24eefaf300f7f84d86b1 except for Kixunil's comment about deprecation.

File: src/util/schnorr.rs (hunk 4 above)
Comment (dr-orlovsky) [reply src/util/schnorr.rs]: done

**Comment by dr-orlovsky** [comment]
@apoelstra updated, needs re-ACK

**Review by sanket1729 — DISMISSED** [review]
ACK 38fdb159b458d4ccc344f19eb64d0613670380d8

**Review by Kixunil — APPROVED** [review]
ACK 38fdb159b458d4ccc344f19eb64d0613670380d8

**Comment by dr-orlovsky** [comment]
Had to re-base after the latest merges; @sanket1729 @Kixunil @apoelstra pls re-ACK

**Review by apoelstra — DISMISSED** [review]
ACK 0d5d0f349c3c80d364aa3908b4e7c22b52a355a6

**Comment by sanket1729** [comment]
Now this one also has conflicts

**Comment by dr-orlovsky** [comment]
Rebased, another re-ACK required, sorry, @Kixunil @apoelstra @sanket1729

**Review by apoelstra — APPROVED** [review]
re-ACK 406f4c403a04d54daee2448ba7e1de3421c1782a

Took me a moment to figure out why you removed the PSBT changes, but it is because PSBT now uses rust-secp key types and is immune to API changes for rust-bitcoin key types.

**Comment by dr-orlovsky** [comment]
@apoelstra which is good, isn't it? :D

**Review by Kixunil — DISMISSED** [review]
ACK 406f4c403a04d54daee2448ba7e1de3421c1782a

**Comment by dr-orlovsky** [comment]
Rebased after latest merges. @apoelstra @Kixunil @sanket1729 please re-ACK

**Review by apoelstra — DISMISSED** [review]
ACK 05f4d0792e5a5fc52cfb191e801b2fcec192c811

**Review by Kixunil — DISMISSED** [review]
ACK 05f4d0792e5a5fc52cfb191e801b2fcec192c811

**Comment by dr-orlovsky** [comment]
Sorry guys, I know you are tired of this - but while we were-reACKing it it had to be rebased once again :( Hopefully for the last time.

**Review by apoelstra — APPROVED** [review]
ACK 8a993e8a589bd458d542446c714b97a4e124e587

**Review by sanket1729 — APPROVED** [review]
utACK 8a993e8a589bd458d542446c714b97a4e124e587

**Pull request merged**